crypto/sleep.c

```diff
@@ -31,7 +31,8 @@ void OSSL_sleep(uint64_t millis)
     unsigned int s = (unsigned int)(millis / 1000);
     unsigned int us = (unsigned int)((millis % 1000) * 1000);
 
-    sleep(s);
+    if (s > 0)
+        sleep(s);
```

Comment:
Can you elaborate here on how this fixes the issue?  I understand that  it avoids a second yield point (which is in and of itself probably beneficial), but I'm not sure how that avoids the hang you reported.  I would expect that calling sleep(0) would result in the thread skipping the cancellation point, and returning immediately from this call.  It seems like you're indicating that a call to sleep(0) yields even though the elapsed time requirement (0) has been met, and another thread never returns it to the cpu.  If thats the case, is there a proscribed method for a single threaded applicaiton to preform a timed wait operation?

Reply:
I do not think this change fixes the issue as much as reduces the number of unnecessary thread switches. There are really multiple situations:

In a kernel thread environment (e.g., Linux), context switching happens entirely based on an interrupt switching strategy, regardless of whether the program itself has threads. sleep may or may not cause invoke a switch attempt. Context switching is more random and related to the overall system load.

In a cooperative threading module (a.k.a. PUT), a switch can happen based on operations being performed (sleep, *printf, yield, etc.) - with results being more sensitive to the number of switches and when they occur. Putting debug messages in the code (running with V=1 vs. not, using a gdb), can (and very often does) influence the outcome, which is what I am seeing. So reducing unnecessary switches can be, but is not always, helpful.

(I am happy to debate the performance merits of each model, but obviously not here).

In unthreaded code, system interrupts and other time-slicing can cause context switch between processes, like perl and openssl, but not within a process state. This provides more stable results because from the code's POV, there is only one thread - although this does impact who waits for what when interprocess communication is involved. There are generally fewer differences in the behaviour between SMP and MPP architectures for unthreaded code, at least these days. Up until around the 1990s, NonStop was entirely cooperative even in the kernel (limited preemptive process swaps), which made life difficult for porting. Changing helped massively for applications on the platform.

My experience is that kernel threading can have more random, but consistent results based on system load, while cooperative depends more on in-process sequence of operations and switch sequencing, with I/O operations heavily impacting the results.

Very well hardened threaded code should run correctly in both models, which is why I do not think the root cause of why the threads appear to have inconsistent behaviour in the cooperative models has been identified sufficiently, though. **Disclaimer:** This statement is more gut feel based on output results, not based on a sufficient analysis of the threading code, so please do not take it personally.

To this change specifically, we had discussed it in prior pull requests/issues, so I tried it, and it did make a difference in `70-test_quic_lcidm` and `70-test_quic_multistream` when testing with `V=1`. I am assuming from differences using `V=1` or not that this results from fewer switches - OSSL_sleep(1000) will cause 1 potential switch, not 2, with this change. This is why I put out the PR, but there are no "Possibly Fixes: #24009" or "Might be Helpful to: #24009" annotations in GitHub comments.

I am not married to this PR, but I cannot approve beta1 in its current state.

Reply:
I'm not arguing that avoiding a sleep(0) isn't more efficient, I agree that it is.  I'm more trying to determine if we've actually fixed the problem reported in https://github.com/openssl/openssl/issues/24009 , Or just hidden it.  I'm trying to understand what the behavior of sleep is on nonstop when using a PUT threading model.  it would seem from your findings here that when sleep is called, the called process/thread yields, and if something doesn't happen to put it back on the cpu, then we're hung.  This obviously fixes the case where OSSL_sleep is called when time < 1000 ms, but if we sleep for more than a 1000 ms, will we find ourselves back in the same boat?

Reply:
The call to sleep never actually hangs. What might happen is this (and it is not specific to NonStop, but is to PUT except for the first step - I have used PUT elsewhere a while back but can't remember what version of what with virtually identical behaviour to NonStop):
1. Thread A is started and does something leading to a yield - this better happen or we are in trouble and I think that may have been the case in the first bug with `usleep(1000000)` being called (as in this did not cause a context switch and mucked up the PUT runtime - I am suspecting that our original issue with SPT may have been related to this).
2. Thread B is started and does something leading to a yield - again, hopefully.
3. Thread A calls `sleep(0)`. It is unclear whether this causes a context swap or not. I do not think it does.
4. Thread B is sitting waiting for something to happen.
5. Thread A calls `usleep(<1000000)` and yields if B is waiting.
6. Thread B does its thing and reaches `usleep()` and results in a swap back to A.
-- so far so good, I think.
7. If `fprintf()` is called, the PUT runtime may yield (or not). It seems like it only yields if there is a mandatory fflush(), because we get a sizable chunks of output but does not necessarily require an LF to trigger a swap. Calling I/O operations causes the thread dispatch behaviour to change sequences, so running with debug messages ends up with different results than without output, making reproducing bugs in debug mode difficult. The optimizer should have an impact in PUT, so `-O0` and `-O2` should behave the same way (from my experience), unless the optimizer changes the timing of when I/O happen between threads. In kernel threading, the optimizer **can** have an impact on thread dispatch sequencing.
8. If Thread A calls `sleep(1)`, it **will** swap context if there is a thread waiting.
9. If we hit the end of program before all threads properly terminate (`pthread_join` or `pthread_exit`), this will create all sorts of potentially random problems. The `atexit()` code has no relationship to PUT whatsoever, so if there openssl has a copy running as  server and is waiting on a `listen()` and has not yet connected when the threaded test process terminates, the openssl (or perl) process hangs waiting - we have seen server processes in this state.

With kernel threads, which is the default in Linux AFAIK, dispatches between threads are done by the kernel, not by the C-runtime in individual processes, so 1,2,7 and 9 above are no factors because threads all will run at least for some period of time. That is not guaranteed in PUT, where yields are not guaranteed except for operations that have implicit yields (anything that maps to `pthread_`_something_, which includes standard I/O operations that hit the PUT runtime).

Reply:
Ok, thank you for that update.  I have a few notes:

referring to the initial report, something is hanging.  It may not be a specific call to OSSL_sleep, but somthing, somewhere, isn't completing.

Based on the initial comment:
```
70-test_quic_lcidm.t .................... 
# The results of this test will end up in test-runs/test_quic_lcidm
1..1
    # Subtest: ../../test/quic_lcidm_test
    1..1
    ok 1 - test_lcidm
../../util/wrap.pl ../../test/quic_lcidm_test => 0
ok 1
ok
```
It doesn't look like its this test thats actually hanging, as it reports an ok output

Further looking at your running process list:
```
 PID TTY      TIME        CMD
 905969852 ?              00:00 make
2097152298 ?              00:00 sleep
 738197525 ?              00:00 sh
1728053513 ?              00:00 make
 100663608 ?              00:05 sh
  16777364 ?              00:00 perl
 134218080 ?              00:00 ../../apps/openssl
1006633143 ?              00:00 sh
1543503948 ?              00:00 make
1845493841 ?              00:00 ../../apps/openssl
 956301601 ?              00:00 ../../apps/openssl
1459618113 ?              15:41 ../../test/quic_multistream_test
 973078851 ?              00:00 ../../apps/openssl
1560281175 ?              00:00 perl
 234881371 ?              00:08 perl
```

It actually looks like its quic_multistream_test thats stuck (I read that as running for 15 minutes and change)

Which makes more sense, as the lcdim test is single threaded, and never makes a call to OSSL_sleep at all (which I understand as you note may not be the source of the problem, but its what this change is related to, so I just want to point that out).  Conversely quic_multistream_test makes use of OSSL_sleep in several locations, so lets focus there.

Looking at the uses of OSSL_sleep, I see two:
A) The server thread calls OSSL_sleep(1) in a for(;;) loop every iteration after tick processing
B) The client script uses OSSL_sleep(op->arg2) in the implementation of its OP_SLEEP script macro, which is used in one location, script_86 in quic_multistream_test, as part of a a loop in which we sleep for 10ms at an interation
C) Based on these, it would seem both threads would wind up calling sleep(0) in each context.  If, as you note in (3) above, a context swap is not generated by this call, That in my mind amounts to a no_op effectively, returning the thread to the cpu immediately.  If that is the case, then I'm struggling to see how avoiding the call to sleep(0), changes anything.  It clearly does something of course, as you note the hang abates, and also as noted, its not a bad thing to introduce just to avoid a needless system call.  but, in an effort to avoid tripping over this problem again down the line, I'd like to understand what the behavior of this call is so we know exactly whats getting fixed.

Reply:
I have a question being opened to the PUT maintainer on how `sleep` behaves in this situation. Hopefully we can get a quick answer.

Reply:
Thank you, hopefully that will resolve this quickly

Reply:
I have not had an answer yet, but found one thing that might be relevant (but I doubt it), since some of the processes have had pending signals during the hangs. From the NonStop man page for sleep (threading):
>If an unblocked signal is received during this suspension period,  thread-aware  `sleep()` returns  control immediately and returns the sleep time remaining.

Reply:
Thats a good note.  I don't suppose there is any way to determine what the pending signal is by any chance?

Reply:
> Thats a good note. I don't suppose there is any way to determine what the pending signal is by any chance?

Only when it happens but not after the fact. I recall seeing two in particular: SIGCHLD and SIGPIPE. Both should interrupt `sleep()`, AFAIK. If these are explicitly ignored, `sleep()` will not be interrupted. SIGALRM will also interrupt `sleep()` but I have not seen it in the pending signal lists for any process - but I might have missed it.